File: cloud-clip/lib/board_column_test.go
package lib

import (
	"encoding/json"
	"fmt"
	"testing"
)

// 看板的列：`POST /content/<id>/column`。
//
// 这是看板的最小实现 —— 固定三列（todo / doing / done）、卡片就是剪贴板条目本身、
// 不建新表也不做列内顺序。这里钉住的是**契约**，不是实现细节：
// 列名归一化、非法输入被拒、错误响应形状统一，以及**挪列不动 timestamp**。
func TestBoardColumnUpdate(t *testing.T) {
	srv := newShortcutServer(t, "", RoomAuthConfig{})

	status, body := shortcutDo(t, "POST", srv.URL+"/text?room=default", "待办事项", "text/plain")
	if status != 200 {
		t.Fatalf("POST /text 失败: %d %s", status, body)
	}
	var created struct {
		ID string `json:"id"`
	}
	if err := json.Unmarshal(body, &created); err != nil {
		t.Fatalf("解析 /text 响应失败: %v (%s)", err, body)
	}

	before := boardContentJSON(t, srv.URL, created.ID)
	if before["column"] != "" {
		t.Fatalf("新条目的列应当是空串（= 待办），实测 %q", before["column"])
	}

	status, body = shortcutDo(t, "POST",
		srv.URL+"/content/"+created.ID+"/column?room=default", `{"column":"doing"}`, "application/json")
	if status != 200 {
		t.Fatalf("挪列失败: %d %s", status, body)
	}
	var moved struct {
		ID     string `json:"id"`
		Type   string `json:"type"`
		Column string `json:"column"`
	}
	if err := json.Unmarshal(body, &moved); err != nil {
		t.Fatalf("解析挪列响应失败: %v (%s)", err, body)
	}
	if moved.Column != "doing" || moved.ID != created.ID || moved.Type != "text" {
		t.Fatalf("挪列响应不对: %+v", moved)
	}

	after := boardContentJSON(t, srv.URL, created.ID)
	if after["column"] != "doing" {
		t.Fatalf("挪列没生效: %v", after["column"])
	}
	// ⚠️ 挪列**不能**改时间戳。updateTextMessage 改正文时会刷时间戳，但挪位置不该让卡片
	// 在时间流里跳到最前面 —— 那等于拖一下就把整个列表重排了。
	if fmt.Sprint(after["timestamp"]) != fmt.Sprint(before["timestamp"]) {
		t.Fatalf("挪列改了时间戳: %v → %v", before["timestamp"], after["timestamp"])
	}

	// 空串归一成 todo（新条目默认落待办，客户端不必自己填）
	status, body = shortcutDo(t, "POST",
		srv.URL+"/content/"+created.ID+"/column?room=default", `{"column":""}`, "application/json")
	if status != 200 {
		t.Fatalf("空串应当被接受并归一成 todo: %d %s", status, body)
	}
	if got := boardContentJSON(t, srv.URL, created.ID)["column"]; got != "todo" {
		t.Fatalf("空串没有归一成 todo: %v", got)
	}

	// 大小写和空白都该容错
	status, _ = shortcutDo(t, "POST",
		srv.URL+"/content/"+created.ID+"/column?room=default", `{"column":"  DONE  "}`, "application/json")
	if status != 200 {
		t.Fatalf("带空白/大写的列名应当容错: %d", status)
	}
	if got := boardContentJSON(t, srv.URL, created.ID)["column"]; got != "done" {
		t.Fatalf("列名没有归一化: %v", got)
	}

	cases := []struct {
		label      string
		method     string
		path       string
		body       string
		wantStatus int
		wantCode   string
	}{
		{"未知列", "POST", "/content/" + created.ID + "/column?room=default", `{"column":"nope"}`, 400, "invalid_column"},
		{"不存在的 id", "POST", "/content/999999/column?room=default", `{"column":"todo"}`, 404, "content_not_found"},
		{"非 POST", "GET", "/content/" + created.ID + "/column?room=default", "", 405, "method_not_allowed"},
		{"请求体不是 JSON", "POST", "/content/" + created.ID + "/column?room=default", "not json", 400, "invalid_body"},
		{"id 不是数字", "POST", "/content/abc/column?room=default", `{"column":"todo"}`, 400, "invalid_content_id"},
	}
	for _, tc := range cases {
		status, body := shortcutDo(t, tc.method, srv.URL+tc.path, tc.body, "application/json")
		if status != tc.wantStatus {
			t.Errorf("%s: 期望 HTTP %d，实测 %d（%s）", tc.label, tc.wantStatus, status, body)
			continue
		}
		boardAssertErrorShape(t, tc.label, body, tc.wantCode)
	}
}

// boardContentJSON 取一条内容的 JSON 形态（顺带验证 column 出现在 HTTP 响应里，
// 不只出现在 WebSocket 载荷里 —— 两处不一致会让「刷新后列丢失」这类问题很难查）。
func boardContentJSON(t *testing.T, base, id string) map[string]interface{} {
	t.Helper()
	status, body := shortcutDo(t, "GET", base+"/content/"+id+"?room=default&json=1", "", "")
	if status != 200 {
		t.Fatalf("GET /content/%s?json=1 失败: %d %s", id, status, body)
	}
	var m map[string]interface{}
	if err := json.Unmarshal(body, &m); err != nil {
		t.Fatalf("解析内容 JSON 失败: %v (%s)", err, body)
	}
	return m
}

// boardAssertErrorShape 错误响应必须恒为 {code,error,message} 且三者都非空
// （与 TestErrorResponseShapeIsStable 同一条契约，这里只覆盖看板新增的错误码）。
func boardAssertErrorShape(t *testing.T, label string, body []byte, wantCode string) {
	t.Helper()
	var e struct {
		Code    string `json:"code"`
		Error   string `json:"error"`
		Message string `json:"message"`
	}
	if err := json.Unmarshal(body, &e); err != nil {
		t.Fatalf("%s: 错误响应不是 JSON: %v (%s)", label, err, body)
	}
	if e.Code != wantCode {
		t.Fatalf("%s: 错误码期望 %q，实测 %q（%s）", label, wantCode, e.Code, body)
	}
	if e.Error == "" || e.Message == "" {
		t.Fatalf("%s: 错误响应必须同时带 error 和 message: %s", label, body)
	}
}

File: cloudflare/workers/test/board-column.test.mjs
// 看板的列：`POST /content/:id/column`。
//
// 与 Go 侧 handleContentColumn 是同一份契约 —— 固定三列（todo / doing / done）、
// 卡片就是剪贴板条目本身、不建新表、**不动 timestamp**。
// 这里钉的是契约，不是实现：列名归一化、错误码、以及「挪列不重排时间流」。
import { ContentHandler } from './.build/content.mjs';
import { TextHandler } from './.build/text.mjs';
import { makeEnv, makeChecker, postJson, getJson } from './harness.mjs';

const { check, summary } = makeChecker();

// harness 的 postJson/getJson 不带 params，而这两个接口要从路径里取 :id ——
// itty-router 平时把 params 塞进 request，测试里得自己挂上（和 auth-room 的 downloadRequest 一样）。
// ⚠️ 凭据也要带：makeEnv 里 AUTH_PASSWORD='123'，不带就是「认证失败」，
// 而报出来是空对象，很容易误读成「接口坏了」。
async function withParams(handler, env, path, { method = 'GET', body = null, params = {} } = {}) {
  const headers = { Authorization: 'Bearer 123' };
  if (body !== null) headers['Content-Type'] = 'application/json';
  const req = Object.assign(
    new Request(`http://worker.local${path}`, { method, headers, body }),
    { params },
  );
  const res = await handler(req, env);
  const text = await res.text();
  let json = null;
  try { json = JSON.parse(text); } catch {}
  return { status: res.status, json, text };
}

async function createText(env, content, room = 'default') {
  const r = await postJson(TextHandler.create, env, `/text?room=${room}`, content);
  return r.json;
}

const setColumn = (env, id, column, opts = {}) => withParams(
  ContentHandler.setColumn, env, `/content/${id}/column?room=${opts.room || 'default'}`,
  { method: 'POST', body: opts.rawBody !== undefined ? opts.rawBody : JSON.stringify({ column }), params: { id: String(id) } },
);

const readContent = (env, id, room = 'default') => withParams(
  ContentHandler.getById, env, `/content/${id}?room=${room}&json=1`, { params: { id: String(id) } },
);

console.log('\n── A. 挪列：生效、落库、且不动时间戳 ──');
{
  const { env } = makeEnv();
  const created = await createText(env, '待办事项');
  const id = created.id;

  const before = (await readContent(env, id)).json;
  check('新条目默认列是空串（= 待办）', before.column, '');

  const res = await setColumn(env, id, 'doing');
  check('挪列返回 200', res.status, 200);
  check('挪列响应带 id', res.json.id, String(id));
  check('挪列响应带 type', res.json.type, 'text');
  check('挪列响应带 column', res.json.column, 'doing');

  const after = (await readContent(env, id)).json;
  check('挪列落库了', after.column, 'doing');
  // ⚠️ 挪列**不能**改时间戳 —— 改了卡片会在时间流里跳到最前面，等于拖一下重排整个列表
  check('挪列不动 timestamp', after.timestamp, before.timestamp);
}

console.log('\n── B. 列名归一化 ──');
{
  const { env } = makeEnv();
  const { id } = await createText(env, '归一化');

  await setColumn(env, id, '');
  check('空串 → todo', (await readContent(env, id)).json.column, 'todo');

  await setColumn(env, id, '  DONE  ');
  check('空白 + 大写 → done', (await readContent(env, id)).json.column, 'done');

  await setColumn(env, id, undefined);
  check('字段缺失 → todo', (await readContent(env, id)).json.column, 'todo');
}

console.log('\n── C. 错误路径（错误码要和 Go 侧逐字一致）──');
{
  const { env } = makeEnv();
  const { id } = await createText(env, '错误路径');

  const bad = await setColumn(env, id, 'nope');
  check('未知列 → 400', bad.status, 400);
  check('未知列 → invalid_column', bad.json.code, 'invalid_column');

  const missing = await setColumn(env, 999999, 'todo');
  check('不存在的 id → 404', missing.status, 404);
  check('不存在的 id → content_not_found', missing.json.code, 'content_not_found');

  const badBody = await setColumn(env, id, null, { rawBody: 'not json' });
  check('请求体不是 JSON → 400', badBody.status, 400);
  check('请求体不是 JSON → invalid_body', badBody.json.code, 'invalid_body');

  const badId = await setColumn(env, 'abc', 'todo');
  check('id 不是数字 → 400', badId.status, 400);
  check('id 不是数字 → invalid_content_id', badId.json.code, 'invalid_content_id');
}

console.log('\n── D. 列要能一路走到前端（刷新后卡片不能全回待办）──');
{
  const { env, db } = makeEnv();
  const { id } = await createText(env, '载荷');
  await setColumn(env, id, 'doing');

  // 库里存的是 boardColumn 列（`column` 是 SQL 关键字，故意不叫那个名字）
  const row = db.prepare('SELECT boardColumn FROM messages WHERE id = ?').get(Number(id));
  check('库里存进 boardColumn 列', row.boardColumn, 'doing');
  // 但对外必须叫 column —— HTTP 载荷和 WebSocket 握手载荷用的是同一个 builder，
  // 只改一处会让「刷新后卡片全回待办」这种问题极难查。
  check('HTTP 载荷映射成 column', (await readContent(env, id)).json.column, 'doing');
}

console.log('\n── E. 空库上也能用（boardColumn 列是自愈加出来的）──');
{
  // makeEnv 刻意停在「没有 boardColumn 列」的旧表结构上，所以这一条真的会走 ALTER
  const { env } = makeEnv();
  const { id } = await createText(env, '自愈加列');
  const res = await setColumn(env, id, 'done');
  check('旧表结构下挪列成功', res.status, 200);
  check('旧表结构下列也落库了', (await readContent(env, id)).json.column, 'done');
}

summary('看板列在 Worker 上成立');
